fiat/prices.go

```diff
@@ -6,16 +6,115 @@ import (
 	"sort"
 	"time"
 
+	"github.com/lightninglabs/faraday/utils"
 	"github.com/lightningnetwork/lnd/lnwire"
 	"github.com/shopspring/decimal"
 )
 
 var (
-	errNoPrices        = errors.New("no price data provided")
-	errPriceOutOfRange = errors.New("timestamp before beginning of price " +
+	errNoPrices            = errors.New("no price data provided")
+	errUnknownPriceBackend = errors.New("unknown price backend")
+	errPriceOutOfRange     = errors.New("timestamp before beginning of price " +
 		"dataset")
+
+	// errGranularityRequired is returned when a request is made that
+	// required fiat prices but the granularity of those prices is not set.
+	errGranularityRequired = errors.New("granularity required when " +
+		"fiat prices are enabled")
+)
+
+// fiatBackend is an interface that must be implemented by any backend that
+// is used to fetch fiat price information.
+type fiatBackend interface {
+	rawPriceData(ctx context.Context, startTime,
+		endTime time.Time) ([]*USDPrice, error)
+}
+
+// PriceSource holds a fiatBackend that can be used to fetch fiat price
+// information.
+type PriceSource struct {
+	impl fiatBackend
+}
+
+// GetPrices fetches price information using the given the PriceSource
+// fiatBackend implementation. GetPrices also validates the time parameters and
+// sorts the results.
+func (p PriceSource) GetPrices(ctx context.Context, startTime,
+	endTime time.Time) ([]*USDPrice, error) {
+
+	// First, check that we have a valid start and end time, and that the
+	// range specified is not in the future.
+	if err := utils.ValidateTimeRange(
+		startTime, endTime, utils.DisallowFutureRange,
+	); err != nil {
+		return nil, err
+	}
+
+	historicalRecords, err := p.impl.rawPriceData(ctx, startTime, endTime)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort by ascending timestamp once we have all of our records. We
+	// expect these records to already be sorted, but we do not trust our
+	// external source to do so (just in case).
+	sort.SliceStable(historicalRecords, func(i, j int) bool {
+		return historicalRecords[i].Timestamp.Before(
+			historicalRecords[j].Timestamp,
+		)
+	})
+
+	return historicalRecords, nil
+}
+
+// PriceBackend is an enum that indicates which backend we are using for fiat
+// information.
+type PriceBackend int
+
+const (
+	UnknownPriceBackend PriceBackend = iota
+
+	// CoinCapPriceBackend uses CoinCap's API for fiat price data.
+	CoinCapPriceBackend
+
+	// CoinDeskPriceBackend uses CoinDesk's API for fiat price data.
+	CoinDeskPriceBackend
 )
 
+var priceBackendNames = map[PriceBackend]string{
+	UnknownPriceBackend:  "",
+	CoinCapPriceBackend:  "coincap",
+	CoinDeskPriceBackend: "coindesk",
+}
+
+func (p PriceBackend) ToString() string {
+	return priceBackendNames[p]
+}
+
```

Comment:
hmm so #117 adds another one of these maps. Could get messy if more things are added. How about something like this rather:

```
type backendsConfig struct {
	name string
	proxySupport bool
}

var priceBackendsConfig = map[PriceBackend]backendsConfig {
	UnknownPriceBackend: {},
	CoinCapPriceBackend: {
		name: "coincap",
		proxySupport: false,
	},
	CoinDeskPriceBackend: {
		name: "coindesk",
		proxySupport: true,
	},
}
```